### PR TITLE
Adding L0 tests for Legacy NodeHandler logic | Adding 2 Knobs for feature flag & EOL toggle status value

### PR DESCRIPTION
**Pull request by rishabhmalikMS**
### **Context**
This PR introduces comprehensive L0 test specifications for NodeHandler behavior to establish a test-driven development foundation for upcoming EOL (End of Life) Node.js version policy and unified node selection strategy improvements. The test specifications define expected behaviors for both current legacy implementation and future unified strategy implementation.

Also added 2 knobs, one for feature flag usage to switch between legacy and new Node selection process. Another, to read the EOL toggle status value injected by server when user changes task settings from org settings from ADO.

**Related Work**: This is the first PR in a multi-part effort to modernize Node.js version handling in Azure Pipelines Agent with proper EOL policy enforcement and custom node path support.

**Workitems links**
- [AB#2340217](https://dev.azure.com/mseng/b924d696-3eae-4116-8443-9a18392d8544/_workitems/edit/2340217)
- [AB#2339023](https://dev.azure.com/mseng/b924d696-3eae-4116-8443-9a18392d8544/_workitems/edit/2339023)
- [AB#2339024](https://dev.azure.com/mseng/b924d696-3eae-4116-8443-9a18392d8544/_workitems/edit/2339024)

---

### **Description**
Added comprehensive test infrastructure and specifications for NodeHandler with detailed test scenarios covering:

**Test Infrastructure**:
- `NodeHandlerTestBase.cs` - Clean base class for NodeHandler test execution with environment setup/teardown
- `NodeHandlerL0.TestSpecifications.cs` -  Comprehensive test scenarios covering all Node.js versions and edge cases
- `NodeHandlerL0.AllSpecs.cs` - Unified test runner executing all scenarios
- `NodeHandlerCollections.cs` - Test collection configuration for proper isolation

**Test Coverage Areas**:
2. **Node6 Scenarios** - Legacy Node 6.x behavior and EOL policy interactions  
3. **Node10 Scenarios** - Node 10.x behavior and EOL policy interactions
4. **Node16 Scenarios** - Node 16.x behavior and EOL policy interactions
5. **Node20 Scenarios** - Current default Node 20.x behavior
6. **Node24 Scenarios** - Latest Node 24.x support
7. **Container Scenarios** - Node selection in containerized environments
8. **EOL Policy Scenarios** - End-of-life version handling differences between legacy and unified strategies

**Test Structure**: Each test scenario includes:
- Input configuration (handler type, environment knobs, glibc errors, container mode)
- Strategy-specific expectations (legacy vs unified behavior)
- Clear success/failure criteria and expected error messages
- Please refer [TestScenario Field Structure & Usage](https://microsoftapc.sharepoint.com/:w:/t/ADOTasksandAgents/IQBFmFZjUQ0IRL1Dg31H05r9AdPDMIL_7VfiTSMhabqzyp8?e=bTM6rO) regarding our test specifications have been defined and used.

Details of each test scenario added/covered: [Test scenarios](https://dev.azure.com/mseng/AzureDevOps/_wiki/wikis/AzureDevOps.wiki/51766/Node.js-Version-Selection-Test-Scenarios)

---

### **Risk Assessment** (Low / Medium / High)  
**Low Risk** - This PR contains only test code with no changes to production logic. The tests currently run against existing legacy NodeHandler implementation, with some CustomNode tests intentionally failing to demonstrate implementation gaps for future development.

---

### **Unit Tests Added or Updated** (Yes / No)  
Yes

---

### **Additional Testing Performed**
_List manual or automated tests performed beyond unit tests (e.g., integration, scenario, regression)._

--- 

### **Change Behind Feature Flag** (Yes / No)
Not applicable - This PR contains only test code with no production changes requiring feature flags.

---

### **Tech Design / Approach**
- Design has been written and reviewed. 
- Any architectural decisions, trade-offs, and alternatives are captured. 

---

### **Documentation Changes Required** (Yes/No)
**Test-First Development Approach**:
- Comprehensive test specifications define behavior before implementation
- Clear separation between legacy and unified strategy expectations
- Strategy-agnostic test infrastructure supporting both approaches

---
### **Logging Added/Updated** (Yes/No)
- Appropriate log statements are added with meaningful messages. 
- Logging does not expose sensitive data. 
- Log levels are used correctly (e.g., info, warn, error). 

--- 

### **Telemetry Added/Updated** (Yes/No) 
- Custom telemetry (e.g., counters, timers, error tracking) is added as needed. 
- Events are tagged with proper metadata for filtering and analysis. 
- Telemetry is validated in staging or test environments.

---

### **Rollback Scenario and Process** (Yes/No)
- Rollback plan is documented. 

---

### **Dependency Impact Assessed and Regression Tested** (Yes/No)
- All impacted internal modules, APIs, services, and third-party libraries are analyzed. 
- Results are reviewed and confirmed to not break existing functionality.

### CODE REVIEW

File: src/Agent.Sdk/Knob/AgentKnobs.cs [hunk 1]
```diff
@@ -568,6 +568,13 @@ public class AgentKnobs
             new EnvironmentKnobSource("AGENT_DISABLE_NODE6_TASKS"),
             new BuiltInDefaultKnobSource("false"));
 
+        public static readonly Knob EnableEOLNodeVersionPolicy = new Knob(
+            nameof(EnableEOLNodeVersionPolicy),
+            "When enabled, automatically upgrades tasks using end-of-life Node.js versions (6, 10, 16) to supported versions (Node 20.1 or Node 24). Throws error if no supported versions are available on the agent.",
+            new PipelineFeatureSource("AGENT_ENABLE_EOL_NODE_VERSION_POLICY"),
+            new EnvironmentKnobSource("AGENT_ENABLE_EOL_NODE_VERSION_POLICY"),
```
Comment (dassayantan24) [inline src/Agent.Sdk/Knob/AgentKnobs.cs]: Can we please finalize the Knob name here? server side change the Knob name is AGENT_RESTRICT_EOL_NODE_VERSIONS.

File: src/Test/L0/NodeHandlerL0.cs [hunk 2]
```diff
@@ -15,6 +15,7 @@
 
 namespace Microsoft.VisualStudio.Services.Agent.Tests
 {
+    [Collection("Unified NodeHandler Tests")]
```
Comment (dassayantan24) [inline src/Test/L0/NodeHandlerL0.cs:18]: Is this intended?

Comment (rishabhmalikMS) [reply src/Test/L0/NodeHandlerL0.cs:18]: Yes, this is intended. 
Reason: For node handler tests, each test requires setting up creatain environment knobs and flags for it. By default, xUnit runs tests in parallel. As a result, the environment settings get leaked from one test to another causing inconsistent test failures. This collection is created to run node handler tests in sequence to prevent this.

File: src/Agent.Sdk/Knob/AgentKnobs.cs (hunk 1 above)
Comment (rishabhmalikMS) [reply src/Agent.Sdk/Knob/AgentKnobs.cs]: Keep you change as is. Will update knob string here in agent code to match: AGENT_RESTRICT_EOL_NODE_VERSIONS

File: src/Test/L0/NodeHandlerL0.TestSpecifications.cs [hunk 3]
```diff
@@ -0,0 +1,892 @@
...
+                description: "Node6 handler works when in default behavior (EOL policy disabled)",
+                handlerData: typeof(NodeHandlerData),
+                knobs: new() {},
+                expectedNode: "node",
+                expectSuccess: true,
+                shouldMatchBetweenModes: true
+            ),
+
+            new TestScenario(
+                name: "Node6_DefaultBehavior_EOLPolicyDisabled",
```
Comment (sanjuyadav24) [inline src/Test/L0/NodeHandlerL0.TestSpecifications.cs:31]: whats the difference between above and this test case?

File: src/Test/L0/NodeHandlerL0.TestSpecifications.cs [hunk 4]
```diff
@@ -0,0 +1,892 @@
...
+                legacyExpectedNode: "node",
+                legacyExpectSuccess: true,
+                unifiedExpectedNode: "node24",
+                unifiedExpectSuccess: true,
+                shouldMatchBetweenModes: false
+            ),
+
+            new TestScenario(
+                name: "Node6_WithGlobalUseNode10Knob",
+                description: "Node6 handler with global Node10 knob: legacy uses Node10, unified ignores deprecated knob and uses Node6",
```
Comment (sanjuyadav24) [inline src/Test/L0/NodeHandlerL0.TestSpecifications.cs]: why node10 doesn't overwrite node6 knob?

Comment (rishabhmalikMS) [reply src/Test/L0/NodeHandlerL0.TestSpecifications.cs]: AGENT_USE_NODE10 overrides to node 10 in older implementation. In new implementation, as discussed in design discussion(s), we are getting away from using this knob in new approach.

File: src/Agent.Sdk/Knob/AgentKnobs.cs (hunk 1 above)
Comment (rishabhmalikMS) [reply src/Agent.Sdk/Knob/AgentKnobs.cs]: This is updated

File: src/Test/L0/NodeHandlerL0.TestSpecifications.cs (hunk 3 above)
Comment (rishabhmalikMS) [reply src/Test/L0/NodeHandlerL0.TestSpecifications.cs:31]: Essentially these are similar scenarios, for default behavior for node 6 when EOL knob is set to false or the knob itself not being available (this however won't happen as we have default value present in agent knobs). Test added just to cover all basis.

**Comment by tarunramsinghani** [comment]
Wrong PR label, use internal instead of enhancement.

**Comment by rishabhmalikMS** [comment]
> Wrong PR label, use internal instead of enhancement.

Updated

File: src/Test/L0/NodeHandlerTestBase.cs [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+                    if (expectations.ExpectSuccess)
+                    {
+                        string actualLocation = nodeHandler.GetNodeLocation(
+                            node20ResultsInGlibCError: scenario.Node20GlibcError,
+                            node24ResultsInGlibCError: scenario.Node24GlibcError,
+                            inContainer: scenario.InContainer);
+
+                        string expectedLocation = GetExpectedNodeLocation(expectations.ExpectedNode, scenario, thc);
+                        Assert.Equal(expectedLocation, actualLocation);
+                    }
```
Comment (MantavyaDh) [inline src/Test/L0/NodeHandlerTestBase.cs]: What does ExpectSuccess mean? We can assert the output of GetNodeLocation with our ExpectedNodeLocation, to handle failure of the function GetNodeLocation we can maybe use a try catch.

File: src/Test/L0/NodeHandlerTestBase.cs [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+                        var exception = Assert.Throws(scenario.ExpectedErrorType ?? typeof(FileNotFoundException),
+                            () => nodeHandler.GetNodeLocation(
+                                node20ResultsInGlibCError: scenario.Node20GlibcError,
+                                node24ResultsInGlibCError: scenario.Node24GlibcError,
+                                inContainer: scenario.InContainer));
+
+                        if (!string.IsNullOrEmpty(expectations.ExpectedError))
+                        {
+                            Assert.Contains(expectations.ExpectedError, exception.Message);
+                        }
```
Comment (MantavyaDh) [inline src/Test/L0/NodeHandlerTestBase.cs]: If we are expecting an Exception to be thrown by function, why don't we catch the exception and there we check if the exception was expected, in this way we will not have to call the function GetNodeLocation twice.

File: src/Agent.Sdk/Knob/AgentKnobs.cs [hunk 7]
```diff
@@ -568,6 +568,13 @@ public class AgentKnobs
             new EnvironmentKnobSource("AGENT_DISABLE_NODE6_TASKS"),
             new BuiltInDefaultKnobSource("false"));
 
+        public static readonly Knob EnableEOLNodeVersionPolicy = new Knob(
+            nameof(EnableEOLNodeVersionPolicy),
+            "When enabled, automatically upgrades tasks using end-of-life Node.js versions (6, 10, 16) to supported versions (Node 20.1 or Node 24). Throws error if no supported versions are available on the agent.",
```
Comment (MantavyaDh) [inline src/Agent.Sdk/Knob/AgentKnobs.cs]: This is slightly confusing, we are not upgrading tasks, rather when flag is enabled we disregard the EOL Node versions a task have.

File: src/Test/L0/NodeHandlerL0.TestSpecifications.cs [hunk 8]
```diff
@@ -0,0 +1,890 @@
...
+
+            new TestScenario(
+                name: "Node6_WithGlobalUseNode10Knob",
+                description: "Node6 handler with global Node10 knob: legacy uses Node10, strategy-based ignores deprecated knob and uses Node6",
+                handlerData: typeof(NodeHandlerData),
+                knobs: new() { ["AGENT_USE_NODE10"] = "true" },
+                legacyExpectedNode: "node10",
+                legacyExpectSuccess: true,
+                strategyExpectedNode: "node",
+                strategyExpectSuccess: true,
```
Comment (sanjuyadav24) [inline src/Test/L0/NodeHandlerL0.TestSpecifications.cs]: why this should use node6?

File: src/Test/L0/NodeHandlerL0.TestSpecifications.cs [hunk 9]
```diff
@@ -0,0 +1,890 @@
...
+                expectedNode: "node",
+                expectSuccess: true,
+                shouldMatchBetweenLegacyAndStrategy: true
+            ),
+
+            new TestScenario(
+                name: "Node6_EOLPolicyEnabled_UpgradesToNode24",
+                description: "Node6 handler with EOL policy: legacy allows Node6, strategy-based upgrades to Node24",
+                handlerData: typeof(NodeHandlerData),
+                knobs: new() { ["AGENT_ENABLE_EOL_NODE_VERSION_POLICY"] = "true" },
```
Comment (raujaiswal) [inline src/Test/L0/NodeHandlerL0.TestSpecifications.cs]: I think after Sayantan suggestion, we have changed to AGENT_RESTRICT_EOL_NODE_VERSIONS in Agent Knob, but forget to change here and multiple places, could you please check if we want to update here as well.

File: src/Test/L0/NodeHandlerTestBase.cs [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+
+                    if (expectations.ExpectSuccess)
+                    {
+                        string actualLocation = nodeHandler.GetNodeLocation(
+                            node20ResultsInGlibCError: scenario.Node20GlibcError,
+                            node24ResultsInGlibCError: scenario.Node24GlibcError,
+                            inContainer: scenario.InContainer);
+
+                        string expectedLocation = GetExpectedNodeLocation(expectations.ExpectedNode, scenario, thc);
+                        Assert.Equal(expectedLocation, actualLocation);
```
Comment (sanjuyadav24) [inline src/Test/L0/NodeHandlerTestBase.cs:87]: in the scenario we had two values, legacyexpected value and strategy expected value, here we are checking only one
is this expected?

File: src/Test/L0/NodeHandlerL0.TestSpecifications.cs (hunk 8 above)
Comment (rishabhmalikMS) [reply src/Test/L0/NodeHandlerL0.TestSpecifications.cs]: Because in new implementation node 10 (useNode10) knob is deprecated.

File: src/Test/L0/NodeHandlerTestBase.cs (hunk 10 above)
Comment (rishabhmalikMS) [reply src/Test/L0/NodeHandlerTestBase.cs:87]: Yes, right now this is expected. Because the tests are running only on existing original code of getNodeLocation. With our strategy code integration both node expectation will be included here.

File: src/Test/L0/NodeHandlerL0.TestSpecifications.cs (hunk 9 above)
Comment (rishabhmalikMS) [reply src/Test/L0/NodeHandlerL0.TestSpecifications.cs]: This is updated

File: src/Agent.Sdk/Knob/AgentKnobs.cs (hunk 7 above)
Comment (rishabhmalikMS) [reply src/Agent.Sdk/Knob/AgentKnobs.cs]: This is updated to be more descriptive

File: src/Test/L0/NodeHandlerTestBase.cs (hunk 6 above)
Comment (rishabhmalikMS) [reply src/Test/L0/NodeHandlerTestBase.cs]: This is updated to use try-catch approach and reduce fields from test scenario simplifying them further

File: src/Test/L0/NodeHandlerTestBase.cs (hunk 5 above)
Comment (rishabhmalikMS) [reply src/Test/L0/NodeHandlerTestBase.cs]: This is removed as per our discussion on below comment

File: src/Test/L0/NodeHandlerL0.TestSpecifications.cs [hunk 11]
```diff
@@ -0,0 +1,802 @@
...
+            string name, 
+            string description,
+            Type handlerData,
+            Dictionary<string, string> knobs = null,
+            string expectedNode = null,
+            string legacyExpectedNode = null,
+            string strategyExpectedNode = null,
+            string strategyExpectedError = null,
+            Type expectedErrorType = null,
+            bool shouldMatchBetweenLegacyAndStrategy = true,
```
Comment (MantavyaDh) [inline src/Test/L0/NodeHandlerL0.TestSpecifications.cs:778]: I think we can remove this now, anyways not being used in this PR.

Comment (rishabhmalikMS) [reply src/Test/L0/NodeHandlerL0.TestSpecifications.cs:778]: We will update this parameter usage with strategy framework code and its test integration.

**Review by sanjuyadav24 — APPROVED** [review]
